asset/asset.go

```diff
@@ -861,6 +862,130 @@ func DeriveGroupKey(genSigner GenesisSigner, genBuilder GenesisTxBuilder,
 	}, nil
 }
 
+// DeriveCustomGroupKey derives an asset's group key based on a signing
+// descriptor, the original group asset genesis, and the asset's genesis.
+func DeriveCustomGroupKey(genSigner GenesisSigner, genBuilder GenesisTxBuilder,
+	signDesc lndclient.SignDescriptor, tapLeaf *psbt.TaprootTapLeafScript,
+	scriptWitness []byte, initialGen Genesis, newAsset *Asset) (*GroupKey,
+	error) {
+
+	// First, perform the final checks on the asset being authorized for
+	// group membership.
+	if newAsset == nil {
+		return nil, fmt.Errorf("grouped asset cannot be nil")
+	}
+
+	if !newAsset.HasGenesisWitness() {
+		return nil, fmt.Errorf("asset is not a genesis asset")
+	}
+
+	if initialGen.Type != newAsset.Type {
+		return nil, fmt.Errorf("asset group type mismatch")
+	}
+
+	// Perform sanity checks on the given signing descriptor. We require a
+	// public key to use as the internal group key.
+	if signDesc.KeyDesc.PubKey == nil {
+		return nil, fmt.Errorf("missing group internal key")
+	}
+
+	if signDesc.DoubleTweak != nil {
+		return nil, fmt.Errorf("bad sign descriptor for group key")
+	}
+
+	// Compute the tweaked group key and set it in the asset before
+	// creating the virtual minting transaction.
+	genesisTweak := initialGen.ID()
+	tweakedGroupKey, err := GroupPubKey(
+		signDesc.KeyDesc.PubKey, genesisTweak[:], signDesc.TapTweak,
```

Comment:
Is the `signDesc` actually needed here if we're already obtaining the witness? IIUC, the last arg here should be the tapscript _root_ for the script group tree.

Reply:
True, I missed that it's converted into the tweak inside `ComputeTaprootOutputKey`.